ref returns and locals have problems with binding  "ref dynamic"

I suspect it could be result of the recent refactoring of ref returns/locals syntax.
Note that   ref dynamic parameters are ok.

```cs

   class Program
    {
        static object[] arr = new object[] { "a" };

        static void Main(string[] args)
        {
            RefParam(ref arr[0]);
            System.Console.WriteLine(x.ToString());


            // error
            ref dynamic x = ref arr[0];
            x = "b";

            System.Console.WriteLine(x.ToString());

            RefReturn() = "e";
            System.Console.WriteLine(x.ToString());

        }

        // works fine
        static void RefParam(ref dynamic p)
        {
            p = "c";
        }

        // error
        static ref dynamic RefReturn()
        {
            return ref arr[0];
        }
    }
```

error CS0246: The type or namespace name 'dynamic' could not be found (are you missing a using directive or an assembly reference?)



NOTE: I tried to make a repro similar to https://github.com/dotnet/roslyn/issues/14166 , but with dynamic, not with tuples.
I think we might have same issue with dynamic and my fix for #14166 would fix it, but that is theoretical, since I can't make "ref dynamic" to work at all in these contexts.

--- a/src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs
+++ b/src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs
@@ -138,6 +138,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                     case RefValueExpression:
                         return ((RefValueExpressionSyntax)parent).Type == node;
 
+                    case RefType:
+                        return ((RefTypeSyntax)parent).Type == node;
+
                     case Parameter:
                         return ((ParameterSyntax)parent).Type == node;
 

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs
@@ -960,5 +960,43 @@ class C
                 Diagnostic(ErrorCode.ERR_BadArgExtraRef, "i").WithArguments("1", "ref").WithLocation(12, 20)
                 );
         }
+
+        [Fact, WorkItem(14174, "https://github.com/dotnet/roslyn/issues/14174")]
+        public void RefDynamicBinding()
+        {
+            var text = @"
+class C
+{
+    static object[] arr = new object[] { ""f"" };
+    static void Main(string[] args)
+    {
+        System.Console.Write(arr[0].ToString());
+
+        RefParam(ref arr[0]);
+        System.Console.Write(arr[0].ToString());
+
+        ref dynamic x = ref arr[0];
+        x = ""o"";
+        System.Console.Write(arr[0].ToString());
+
+        RefReturn() = ""g"";
+        System.Console.Write(arr[0].ToString());
+    }
+
+    static void RefParam(ref dynamic p)
+    {
+        p = ""r"";
+    }
+
+    static ref dynamic RefReturn()
+    {
+        return ref arr[0];
+    }
+}
+";
+            CompileAndVerify(text,
+                expectedOutput: "frog",
+                additionalRefs: new[] { SystemCoreRef, CSharpRef }).VerifyDiagnostics();
+        }
     }
 }

